### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by davies**

### CODE REVIEW

**Comment by mengxr** [comment]
@davies I tried to generate the doc but it doesn't "compile":

~~~
/Users/meng/src/spark/python/docs/pyspark.ml.rst:10: WARNING: autodoc: failed to import module u'pyspark.ml'; the following exception was raised:
Traceback (most recent call last):
  File "/Users/meng/anaconda/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 335, in import_object
    __import__(self.modname)
  File "/Users/meng/src/spark/python/pyspark/ml/__init__.py", line 18, in <module>
    from pyspark.ml.param import *
  File "/Users/meng/src/spark/python/pyspark/ml/param/__init__.py", line 20, in <module>
    from pyspark.ml.util import Identifiable
  File "/Users/meng/src/spark/python/pyspark/ml/util.py", line 23, in <module>
    from pyspark.ml.param import Params
ImportError: cannot import name Params
~~~

File: python/pyspark/sql.py [hunk 1]
```diff
@@ -1794,20 +1794,6 @@ def __repr__(self):
             return "<Row(%s)>" % ", ".join(self)
 
 
-def inherit_doc(cls):
```
Comment (mengxr) [inline python/pyspark/sql.py]: Does SQL still need this function?

Comment (davies) [inline python/pyspark/sql.py]: No

**Comment by davies** [comment]
Hold on, i'm still work on it.

**Comment by davies** [comment]
@mengxr The docs and tests are ready, please review!

File: python/pyspark/ml/param/__init__.py [hunk 2]
```diff
@@ -15,11 +15,24 @@
...
 
+import uuid
 from abc import ABCMeta
 
-from pyspark.ml.util import Identifiable
+__all__ = ['Param', 'Params']
 
-__all__ = ["Param"]
+
+class Identifiable(object):
```
Comment (mengxr) [inline python/pyspark/ml/param/__init__.py]: Should we put `Identifiable` under `param`? I think it should either under `ml` or `ml.util`.

**Pull request merged**